### ISSUE TITLE
Refactor cross-boundary error handling

### ISSUE THREAD

**Issue by neilalexander** [matrix-org/dendrite#1441]
Errors in Dendrite are a mess.

The client API has `jsonerror.MatrixError`, which is nice because those are the things that are returned to clients, although it is maybe misplaced since we also might want to use those in the federation API.

[gomatrix](https://github.com/matrix-org/gomatrix) also has `*gomatrix.HTTPError`, and then in some places we have `roomserverAPI.PerformError`, and in other places we just throw regular opaque errors around. In some API functions we put `PerformError`s into the response struct, in others we just return an `error` from the function itself.

This makes it really difficult to return meaningful clients and other federated servers to errors consistently in any place that an API query is performed.

We should do the work to ensure that if we provide a `jsonerror.MatrixError`, that it is correctly marshalled and unmarshalled across API boundaries and that they make it all the way back to clients/federation endpoints, and tidy up the rest of the mess so that it's consistent and usable.

**Comment by ara4n** [comment]
I agree this is super desirable, but i'm not sure it should technically block entering beta.

**Comment by Kegsay** [comment]
> that it is correctly marshalled and unmarshalled across API boundaries and that they make it all the way back to clients/federation endpoints

iirc this broadly does happen, and there's sytests to prove it. The internal APIs are a mixture of error return / error struct though which should definitely be fixed. Agreed it shouldn't block beta.
